### PR TITLE
check if column in spearman rho is entirely NaN or Inf

### PR DESCRIPTION
**Pull request by kristianeschenburg**
#### Reference issue

Closes Issue [12458](https://github.com/scipy/scipy/issues/12458)

#### What does this implement/fix?

Running current version of ```spearmanr``` with an entire column of NaN or Inf, using nan_policy='omit', results in a raised ```ValueError```.  This error is produced by ```mstats_basic.spearmanr._spearmanr_2cols```, after masking the numpy array.  The masked array (with a constant NaN or Inf column) is empty with shape=(0,2), and raises a ```ValueError``` in the next few lines.

The code now checks for the empty masked array and returns ```SpearmanResult(np.nan, np.nan)```, as would be expected when using nan_policy='omit'.

### CODE REVIEW

File: scipy/stats/mstats_basic.py [hunk 1]
```diff
@@ -503,6 +503,10 @@ def _spearmanr_2cols(x):
         x = ma.mask_rowcols(x, axis=0)
         x = x[~x.mask.any(axis=1), :]
 
+        # If either column is entirely NaN or Inf
```
Comment (rlucas7) [inline scipy/stats/mstats_basic.py:506]: Thanks for the PR. Please add a unit test. The example in the [associated issue](https://github.com/scipy/scipy/issues/12458) (wrapped appropriately in the scipy testing harness) would work.

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@kristianeschenburg thanks for reporting the issue and cutting the PR, needs a unit test with the PR.

**Comment by mdhaber** [comment]
I took the liberty of adding it. Thanks @kristianeschenburg!

File: scipy/stats/mstats_basic.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/mstats_basic.py:506]: I forgot to add -> @kristianeschenburg

File: scipy/stats/tests/test_stats.py [hunk 2]
```diff
@@ -581,6 +581,19 @@ def test_nan_policy(self):
         assert_raises(ValueError, stats.spearmanr, x, x, nan_policy='raise')
         assert_raises(ValueError, stats.spearmanr, x, x, nan_policy='foobar')
 
+    def test_nan_policy_bug_12458(self):
+        np.random.seed(5)
+        x = np.random.rand(5, 10)
```
Comment (rlucas7) [inline scipy/stats/tests/test_stats.py:586]:
```suggestion
        x = np.random.default_rng(5).uniform(0, 1, 50).reshape(5,10)
```
reason: numpy docs indicate `numpy.random.rand()` is deprecated. 

The proposal achieves the same thing. 
IIRC the deprecation of `.rand()` is (somehow) related to the switch the the `Generator` for generating random variates. If this is incorrect please let me know.

File: scipy/stats/tests/test_stats.py [hunk 3]
```diff
@@ -581,6 +581,19 @@ def test_nan_policy(self):
         assert_raises(ValueError, stats.spearmanr, x, x, nan_policy='raise')
         assert_raises(ValueError, stats.spearmanr, x, x, nan_policy='foobar')
 
+    def test_nan_policy_bug_12458(self):
+        np.random.seed(5)
```
Comment (rlucas7) [inline scipy/stats/tests/test_stats.py:585]: If you agree w/suggestion proposed beneath then the `np.random.seed` line can be deleted.

File: scipy/stats/tests/test_stats.py [hunk 4]
```diff
@@ -581,6 +581,19 @@ def test_nan_policy(self):
...
+        x = np.random.rand(5, 10)
+        k = 6
+        x[:, k] = np.nan
+        y = np.delete(x, k, axis=1)
+        corx, px = stats.spearmanr(x, nan_policy='omit')
+        cory, py = stats.spearmanr(y)
+        corx = np.delete(np.delete(corx, k, axis=1), k, axis=0)
+        px = np.delete(np.delete(px, k, axis=1), k, axis=0)
+        assert_allclose(corx, cory)
+        assert_allclose(px, py, atol=1e-14)
```
Comment (rlucas7) [inline scipy/stats/tests/test_stats.py:595]: It looks like these 2 lines are comparing differently, `px, py` are compared absolutely whereas `corx, cory` are compared relatively. Is there a reason to not make both comparisons on absolute terms?

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@mdhaber I left of couple (minor) comments, let me know if you agree or not and whether you've got time to update. If you agree with suggestions and don't have time, I can make the changes to the PR. Once we're in agreement on those 2 comments, I'm ready to merge.

File: scipy/stats/tests/test_stats.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:586]: Yeah I think `Generator` doesn't have a `rand` method.

File: scipy/stats/tests/test_stats.py (hunk 3 above)
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:585]:
```suggestion
```

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -581,6 +581,19 @@ def test_nan_policy(self):
...
+        np.random.seed(5)
+        x = np.random.rand(5, 10)
+        k = 6
+        x[:, k] = np.nan
+        y = np.delete(x, k, axis=1)
+        corx, px = stats.spearmanr(x, nan_policy='omit')
+        cory, py = stats.spearmanr(y)
+        corx = np.delete(np.delete(corx, k, axis=1), k, axis=0)
+        px = np.delete(np.delete(px, k, axis=1), k, axis=0)
+        assert_allclose(corx, cory)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:594]:
```suggestion
        assert_allclose(corx, cory, atol=1e-14)
```

File: scipy/stats/tests/test_stats.py (hunk 4 above)
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:595]: I don't really remember, but let's try it.

File: scipy/stats/tests/test_stats.py (hunk 2 above)
Comment (rlucas7) [reply scipy/stats/tests/test_stats.py]: Hmm, maybe I was too quick on this, are we still supporting old versions of numpy that didn't have the `default_rng` method inside random? 

e.g.
https://travis-ci.org/github/scipy/scipy/jobs/726677838#L1314

looks like that's causing the build to fail. I guess we can back out this change to get the CI to pass and use the `random.rand()` approach for now.

File: scipy/stats/tests/test_stats.py [hunk 6]
```diff
@@ -581,6 +581,18 @@ def test_nan_policy(self):
         assert_raises(ValueError, stats.spearmanr, x, x, nan_policy='raise')
         assert_raises(ValueError, stats.spearmanr, x, x, nan_policy='foobar')
 
+    def test_nan_policy_bug_12458(self):
+        x = np.random.default_rng(5).uniform(0, 1, 50).reshape(5,10)
```
Comment (rlucas7) [inline scipy/stats/tests/test_stats.py:585]:
```suggestion
        np.random.seed(5)
        x = np.random.rand(5, 10)
```

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@mdhaber looks like maybe I was too quick on the generator, seems we still need to support older versions of numpy that might not have. 

Unless we need to update the build version of numpy. 

EDIT: The change was effectively just reverted the `rand` stuff, seems with the github suggestion I can commit, so I self committed to test the change. I'll check back tomorrow to see if everything is green, if so I'll squash and merge.

**Comment by mdhaber** [comment]
Thanks @rlucas7

**Comment by rlucas7** [comment]
Everything was green after last nights 'revert'. 
Thanks @kristianeschenburg @pvanmulbregt @mdhaber

**Pull request merged**